### PR TITLE
[FLINK-31626] HsSubpartitionFileReaderImpl should recycle skipped read buffers.

### PR DESCRIPTION
**Pull request by reswqa**
## What is the purpose of the change

*In [FLINK-30189](https://issues.apache.org/jira/browse/FLINK-30189), `HsSubpartitionFileReaderImpl` will skip the buffer has been consumed from memory to avoid double-consumption. But these buffers were not returned to the `BatchShuffleReadBufferPool`, resulting in read buffer leaks. In addition, all loaded buffers should also be recycled after data view released.*

## Brief change log

  - *`HsSubpartitionFileReader` should not read buffers from disk after it has been released.*
  - *`HsSubpartitionFileReaderImpl` should recycle skipped read buffers.*

## Verifying this change

This change added unit tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsSubpartitionConsumerMemoryDataManager.java
Comment (TanYuxin-tyx) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsSubpartitionConsumerMemoryDataManager.java:95]: The method annotations should also add the new argument description `buffersToRecycle`.

**Review by TanYuxin-tyx — COMMENTED** [review]
@reswqa Thanks for fixing the bug of leaking read buffers. 
The change looks good to me overall. I only left some minor comments, thanks.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsSubpartitionFileReaderImpl.java
Comment (reswqa) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsSubpartitionFileReaderImpl.java:281]: Yes, It is for the purpose of deadlocking.

**Comment by reswqa** [comment]
Thanks @TanYuxin-tyx for the review, I have addressed all the comments in the fix-up commit, please take a look again.

**Review by TanYuxin-tyx — APPROVED** [review]
Thanks @reswqa, LGTM after the fixup.

**Comment by reswqa** [comment]
Rebased it and waiting for CI passing.

**Comment by reswqa** [comment]
merged via 59462197.

**Pull request closed without merging**
